### PR TITLE
feat: precise filter for mito parquet reader

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

### Result

Up to 50% latency cut in tsbs scenario

| test                     | main(ms)   | filter(ms) | opt. rate |
|--------------------------|------------|------------|-----------|
|   **cpu-max-all-1***          |   18.20    |   14.13    |   **0.78**    |
|   cpu-max-all-8          |   81.57    |   79.57    |   0.98    |
|   **double-groupby-1***       |   749.72   |   728.76   |   0.97    |
|   double-groupby-5       |   1156.25  |   1193.33  |   1.03    |
|   double-groupby-all     |   1925.41  |   1902.98  |   0.99    |
|   groupby-orderby-limit  |   453.77   |   499.06   |   1.10    |
|   **high-cpu-1***             |   19.01    |   9.69     |   **0.51**    |
|   high-cpu-all           |   6189.08  |   6402.76  |   1.03    |
|   lastpoint              |   6179.84  |   6141.70  |   0.99    |
|   **single-groupby-1-1-1***   |   11.91    |   10.51    |   **0.88**    |
|   **single-groupby-1-1-12***  |   13.97    |   8.58     |   **0.61**    |
|   single-groupby-1-8-1   |   22.24    |   20.39    |   0.92    |
|   **single-groupby-5-1-1***   |   10.29    |   9.26     |   **0.90**    |
|   **single-groupby-5-1-12***  |   15.30    |   10.00    |   **0.65**    |
|   single-groupby-5-8-1   |   30.01    |   30.37    |   1.01    |

**are featured cases*

### Explain

This patch adds a `SimpleFilterEvaluator` struct that can calculate filter expr in the simple form. See its doc for details.

By using `SimpleFilterEvaluator` in parquet reader, it now can filter out invalid rows in advance. The phase it takes effort is right after a batch is retrieved from parquet file. Thus can reduce the consumption in later process phases like sort / dedup etc.

This also fixes a possible bug when an empty batch is read from parquet file, see UT case `test_read_empty_batch`.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/mito2/src/read.rs [hunk 1]
```diff
@@ -653,6 +672,7 @@ impl BatchBuilder {
 
         Ok(Batch {
             primary_key: self.primary_key,
+            pk_values: None,
```
Comment (zhongzc) [inline src/mito2/src/read.rs:675]: Can this value also be set on the write path? So that the inverted index can skip a decode operation.

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/mito2/src/sst/parquet/reader.rs [hunk 2]
```diff
@@ -467,6 +501,9 @@ impl Drop for ParquetReader {
         READ_ROW_GROUPS_TOTAL
             .with_label_values(&["min_max_index_selected"])
             .inc_by(self.metrics.num_row_groups_min_max_selected as u64);
+        PRECISE_FILTER_ROWS_TOTAL
+            .with_label_values(&["parquet"])
+            .inc_by(self.metrics.num_rows_precise_filtered as u64);
```
Comment (zhongzc) [inline src/mito2/src/sst/parquet/reader.rs:506]: where do you set `self.metrics.num_rows_precise_filtered`?

File: src/common/recordbatch/src/filter.rs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    pub fn evaluate_scalar(&self, input: &ScalarValue) -> Result<bool> {
+        let result = self.evaluate_datum(&input.to_scalar())?;
+        Ok(result.value(0))
+    }
+
+    pub fn evaluate_array(&self, input: &ArrayRef) -> Result<BooleanBuffer> {
+        self.evaluate_datum(input)
+    }
+
+    pub fn evaluate_vector(&self, input: VectorRef) -> Result<BooleanBuffer> {
```
Comment (zhongzc) [inline src/common/recordbatch/src/filter.rs:96]:
```suggestion
    pub fn evaluate_vector(&self, input: &VectorRef) -> Result<BooleanBuffer> {
```

File: src/mito2/src/sst/parquet/reader.rs [hunk 4]
```diff
@@ -515,6 +552,100 @@ impl ParquetReader {
...
+            .read_format
+            .metadata()
+            .field_columns()
+            .enumerate()
+            .map(|(i, f)| (f.column_id, i))
+            .collect::<HashMap<_, _>>();
+
+        // Run filter one by one and combine them result
+        // TODO(ruihang): run primary key filter first. It may short circuit other filters
+        for filter in &self.predicate {
```
Comment (zhongzc) [inline src/mito2/src/sst/parquet/reader.rs:596]: if the mask is all false, we can break it early

File: src/common/recordbatch/src/filter.rs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+                    _ => return None,
+                }
+
+                // swap the expr if it is in the form of `literal` `op` `col`
+                let mut op = binary.op;
+                let (lhs, rhs) = match (&*binary.left, &*binary.right) {
+                    (Expr::Column(ref col), Expr::Literal(ref lit)) => (col, lit),
+                    (Expr::Literal(ref lit), Expr::Column(ref col)) => {
+                        // safety: The previous check ensures the operator is able to swap.
+                        op = op.swap().unwrap();
```
Comment (zhongzc) [inline src/common/recordbatch/src/filter.rs:66]: master of datafusion.

File: src/mito2/src/sst/parquet/reader.rs [hunk 6]
```diff
@@ -515,6 +552,100 @@ impl ParquetReader {
...
+    /// Prunes batches by the pushed down predicate.
+    fn prune_batches(&mut self) -> Result<()> {
+        // fast path
+        if self.predicate.is_empty() {
+            return Ok(());
+        }
+
+        let mut new_batches = VecDeque::new();
+        let mut batches = std::mem::take(&mut self.batches);
+        for batch in batches.drain(..) {
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/reader.rs:564]: The following two patterns should work.

```suggestion
        for batch in self.batches.drain(..) {
```

or
```suggestion
        let mut batches = std::mem::take(&mut self.batches);
        for batch in self.batches {
```

File: src/mito2/src/sst/parquet/reader.rs [hunk 7]
```diff
@@ -515,6 +552,100 @@ impl ParquetReader {
...
+    fn precise_filter(&self, mut input: Batch) -> Result<Option<Batch>> {
+        let mut mask = BooleanBuffer::new_set(input.num_rows());
+
+        let field_id_map = self
+            .read_format
+            .metadata()
+            .field_columns()
+            .enumerate()
+            .map(|(i, f)| (f.column_id, i))
+            .collect::<HashMap<_, _>>();
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/reader.rs:592]: Maybe we should maintain this map in the `ReadFormat`

File: src/mito2/src/sst/parquet/reader.rs [hunk 8]
```diff
@@ -515,6 +552,100 @@ impl ParquetReader {
...
+
+        Ok(())
+    }
+
+    /// TRY THE BEST to perform pushed down predicate precisely on the input batch.
+    ///
+    /// Supported filter expr type is defined in [SimpleFilterEvaluator].
+    ///
+    /// When a filter is referencing primary key column, this method will decode
+    /// the primary key and put it into the batch.
```
Comment (killme2008) [inline src/mito2/src/sst/parquet/reader.rs:582]: Document the return value.

File: src/mito2/src/sst/parquet/reader.rs (hunk 2 above)
Comment (waynexia) [reply src/mito2/src/sst/parquet/reader.rs:506]: Good catch... 🫣

File: src/mito2/src/sst/parquet/format.rs [hunk 9]
```diff
@@ -118,8 +118,14 @@ pub(crate) struct ReadFormat {
...
     arrow_schema: SchemaRef,
-    // Field column id to its index in `schema` (SST schema).
+    /// Field column id to its index in `schema` (SST schema).
+    /// In SST schema, fields are stored in the front of the schema.
     field_id_to_index: HashMap<ColumnId, usize>,
+    /// Field column id to their index in the projected schema (
+    /// the schema of [Batch]).
+    ///
+    /// This field is set at the first call to [convert_record_batch](Self::convert_record_batch).
+    field_name_to_projected_index: Option<HashMap<ColumnId, usize>>,
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/format.rs:128]:
```suggestion
    field_id_to_projected_index: Option<HashMap<ColumnId, usize>>,
```

File: src/mito2/src/sst/parquet/format.rs [hunk 10]
```diff
@@ -259,6 +270,20 @@ impl ReadFormat {
         Ok(())
     }
 
+    fn init_name_to_projected_index(&mut self, record_batch: &RecordBatch) -> Result<()> {
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/format.rs:273]:
```suggestion
    fn init_id_to_projected_index(&mut self, record_batch: &RecordBatch) -> Result<()> {
```

File: src/mito2/src/sst/parquet/reader.rs (hunk 4 above)
Comment (waynexia) [reply src/mito2/src/sst/parquet/reader.rs:597]: We don't have a cheap way to check this :smiling_face_with_tear:

File: src/mito2/src/sst/parquet/format.rs [hunk 11]
```diff
@@ -259,6 +270,20 @@ impl ReadFormat {
         Ok(())
     }
 
+    fn init_id_to_projected_index(&mut self, record_batch: &RecordBatch) -> Result<()> {
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/format.rs:273]: Do we need to return a result? Seems it never returns an error.

File: src/mito2/src/sst/parquet/reader.rs [hunk 12]
```diff
@@ -515,6 +552,105 @@ impl ParquetReader {
...
+        // TODO(ruihang): run primary key filter first. It may short circuit other filters
+        for filter in &self.predicate {
+            let column_name = filter.column_name();
+            let column_metadata = self
+                .read_format
+                .metadata()
+                .column_by_name(column_name)
+                .with_context(|| ColumnNotFoundSnafu {
+                    column: column_name,
+                })?;
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/reader.rs:605]: Should we return an error here? If we add a new column, an old SST might not have this column.

**Review by evenyag — APPROVED** [review]

**Pull request merged**